### ISSUE TITLE
Make maps with HUCs

### ISSUE THREAD

**Issue by ldecicco-USGS** [USGS-R/wateRuse#72]

**Comment by dblodgett-usgs** [comment]
Will just put in the latest HUC08s and deal with the consequences in previous compilation years and/or add needed additional HUC08s

**Comment by dblodgett-usgs** [comment]
I added a heavily simplified version of the HUC08s to sysdata.rda with the variable name huc08Poly. @rwdudley-usgs are you the right person to get that wired into the choropleth function and shiny?

**Comment by rwdudley-usgs** [comment]
I'll take a crack at it given it's so closely related to the county-related mapping I worked on. 

On the topic of Shiny, I'm having difficulty getting it to work since the last update. Anyone else? I'm getting errors related to 'no data' and when I try to load any data in it freezes up. I'm running it through the Chrome browser. I had no problems using it until I updated this morning...

**Comment by mamaupin-usgs** [comment]
I had that same problem the second time I launched Shiny.  The first time
in all functions were plotting.  Haven't tried yet this morning.

On Wed, Jun 1, 2016 at 7:02 AM, Rob Dudley notifications@github.com wrote:

> I'll take a crack at it given it's so closely related to the
> county-related mapping I worked on.
> 
> On the topic of Shiny, I'm having difficulty getting it to work since the
> last update. Anyone else? I'm getting errors related to 'no data' and when
> I try to load any data in it freezes up. I'm running it through the Chrome
> browser. I had no problems using it until I updated this morning...
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-222985141, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLb6xNa-jiryVDQ-CBF9L5fsWqu4hfks5qHYLXgaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by ldecicco-USGS** [comment]
Yup, there was a request to take the initial data off the shiny app, so that the user has to choose their own file. Therefore, all the graphs will initially show an error.

**Comment by ldecicco-USGS** [comment]
As for freezing up, that I haven't seen. @rwdudley-usgs  can you send me the data file you are using? Opening data from the internet app is slow for sure, I highly recommend using the app from R. But, of the data sets that I have, I have not had any crash on opening. Also, are you using Firefox or Chrome?

**Comment by rwdudley-usgs** [comment]
I noticed the data were no longer automatically loaded, but I'm having difficulty opening any data files (I've only tried a couple). I'll keep messing around with it - I'll either figure it out or I'll get you more specifics...  BTW I'm running the app from R then pressing the button to open in browser (Chrome).

On a related note, the `wateRuse` package needs to have some built in data for the vignettes to work, and that has not changed, correct? It's only the Shiny app that now defaults to having no data loaded - right?

**Comment by ldecicco-USGS** [comment]
Yeah, the sample data is still in the package for the vignettes and examples. Definitely send me any files that aren't working, I think all the sample data that I have does work, so it would be good to see more examples that don't.

**Comment by jshourds-usgs** [comment]
Just to chime in I was freezing up bigtime last night using the app in R and opening in Browser (default on that machine is Chrome) trying to load in the county dump file in both the App and Browser.  Today everything is working better (this machine defaults to IE though) and I'm loading the same county dump file.

**Comment by rwdudley-usgs** [comment]
The only data I have are the data that were placed on the Google drive that we don't want on Git; I downloaded those data to my local machine. If I go to the `Regional_dump` folder, there are a bunch of files `acounty_xx.txt` where `xx` is the 2-letter data element code. If I load a file with a 2-letter code that doesn't match what is populated in the dropdown in the app, then it freezes or closes. If I open the file that matches the code, it works, but I only have one data element to work with. However, if I shift-select all 19 files to load them all in, it goes through the load process then freezes (using Chrome) or closes (using the R-generated window). I've restarted R, and done a new pull and rebuild and still have these issues.

**Comment by ldecicco-USGS** [comment]
Oh, OK. I've only written the code to open those type of files in a zip, so try (for now) opening the single file:
allcountydata_201604281345414.zip
or
allhuc8data_20160428121053.zip

This also works individually:
wu_countydata_20160415.txt

or, you can open a set of Excel files, like:
Export_1985, Idaho_County.xlsx
Export_1990, Idaho_County.xlsx
etc.

The code will have to be modified to accept a set of text files.

**Comment by rwdudley-usgs** [comment]
Ok. I see. That clears things up. Thank you!

**Comment by jshourds-usgs** [comment]
So this may be an issue with the regional dump file but Total Pop doesn't work for all states using the allcountydata zip file for the choropleths.  I wonder if it is an aging status issue in the regional files? 

![ky](https://cloud.githubusercontent.com/assets/18489683/15715521/bd21776c-27ec-11e6-9bb3-10e1d5892d88.JPG)

![nj](https://cloud.githubusercontent.com/assets/18489683/15715526/c48d523c-27ec-11e6-8b6d-15fac50e53ea.JPG)

**Comment by rwdudley-usgs** [comment]
@dblodgett-usgs I'm wondering how to subset the huc8 data in a similar fashion as the histCounty data. For the county data we did:
`hc.sub <- histCounties[which(histCounties@data[,area.column] == areas),]`
which yielded a Formal Class SpatialPolygonsDataFrame.
I can't use the same syntax with huc8 data. If I do I get the error
`Error in which(huc8@data[, area.column] == areas) : 
  trying to get slot "data" from an object (class "tbl_df") that is not an S4 object`

I can subset the huc8 data if I drop the `@data`, but then it's not a Formal Class SpatialPolygonsDataFrame and I can't make a map out of it. I need some help with properly subsetting the huc8 data.

**Comment by dblodgett-usgs** [comment]
The data is in huc08Poly. huc8 is the lookup table from AWUDS. Maybe I didn't do everything I need to in order to make the data available from within the package? Yeah, looks like I didn't... another PR coming to make the data visible.

**Comment by rwdudley-usgs** [comment]
For implementing HUC8 choropleths into Shiny, how should users interact with it? Right now, the `choropleth_plot.R` function produces a county map for a State of interest. I added a parameter `unit.type` to `choropleth_plot.R` to specify whether to go fetch the county polygons or the huc polygons. For the HUC8 maps, I was thinking we might still use State to sub-select the HUCs that go with that state. But perhaps users would want to select by HUC4 or HUC6? And we also might want to look ahead to the possibility of aquifers being added to the `choropleth_plot.R` function as we consider design. How would selection happen with aquifers? Probably not by State...

For the purposes of getting it functioning for June 17th, perhaps we simply go with selection by State for now, with a check box to switch between counties and hucs? And worry about aquifers and other complexities in later versions?
Thoughts?

**Comment by mamaupin-usgs** [comment]
I agree.  County choropleths can be rendered more easily using State
selections.  There are some aquifers that we don't have digital maps for
(just a few).

Users (I think) will be interested in State-level data (easy for counties
that don't cross State lines).  If we were able to do a HUC rendering of
the data, would it add up the HUC pieces for those that cross state lines
or does hit just show the HUC pieces within the selected State?  Same
question for principal aquifers.

County-aquifer maps could be based on State selections.

On Fri, Jun 3, 2016 at 12:42 PM, Rob Dudley notifications@github.com
wrote:

> For implementing HUC8 choropleths into Shiny, how should users interact
> with it? Right now, the choropleth_plot.R function produces a county map
> for a State of interest. I added a parameter unit.type to
> choropleth_plot.R to specify whether to go fetch the county polygons or
> the huc polygons. For the HUC8 maps, I was thinking we might still use
> State to sub-select the HUCs that go with that state. But perhaps users
> would want to select by HUC4 or HUC6? And we also might want to look ahead
> to the possibility of aquifers being added to the choropleth_plot.R
> function as we consider design. How would selection happen with aquifers?
> Probably not by State...
> 
> For the purposes of getting it functioning for June 17th, perhaps we
> simply go with selection by State for now, with a check box to switch
> between counties and hucs? And worry about aquifers and other complexities
> in later versions?
> Thoughts?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-223660408, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLbzLwrKoUByyRqpdZjjd2gn6QRV3sks5qIHWygaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by rwdudley-usgs** [comment]
We could make the selection such that you get all HUCs that have any area whatsoever in a given State. Sound ok? 
I think if you have no data for certain HUC for your state it would probably plot gray.

**Comment by mamaupin-usgs** [comment]
so, selection criteria is based on states.  ok
and you only get to choose one state?  ok
if huc data in state is missing--blank.
if you choose more than one state, how would split hucs with missing data
on one state get mapped? rather not go there

On Fri, Jun 3, 2016 at 1:21 PM, Rob Dudley notifications@github.com wrote:

> We could make the selection such that you get all HUCs that have any area
> whatsoever in a given State. Sound ok?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-223669872, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLbwIRvxbrbXlH7i1TYhv4tsRp0byRks5qIH6jgaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by rwdudley-usgs** [comment]
Ok that helps narrow down what I'll attempt, thanks.
Perhaps we can deal with multi-state as an enhancement sometime after June 17.

**Comment by mamaupin-usgs** [comment]
yes, that's what a RWUS user will want to be able to do

On Fri, Jun 3, 2016 at 2:23 PM, Rob Dudley notifications@github.com wrote:

> Ok that helps narrow down what I'll attempt, thanks.
> Perhaps we can deal with multi-state as an enhancement sometime after June
> 17.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-223684872, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLb0L2398oiI8OOPj_UUQMrgjCU85Qks5qII1UgaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by rwdudley-usgs** [comment]
I've got `choropleth_plot.R` producing HUC maps using Internal Only data for Idaho:
![image](https://cloud.githubusercontent.com/assets/18488915/15793249/c348cd96-29b0-11e6-9da7-db1b0f0e8f44.png)
And Maine:
![image](https://cloud.githubusercontent.com/assets/18488915/15793284/e6ea99f0-29b0-11e6-8315-9691799a2a63.png)
Can't get the maps to render in Shiny using the same data. Any help @ldecicco-USGS would be appreciated.

**Comment by ldecicco-USGS** [comment]
I think it's working in the app now (once PR#115 is merged). You've got to pick the right state (if you are using the Excel files), and you've got to correctly choose the checkbox (whether it's HUC's or not). We could try to make that smarter, but for now I think it's OK.

@rwdudley-usgs have a look, and if it seems OK, close this issue.

**Comment by rwdudley-usgs** [comment]
Thanks!  I'll check it out.
I felt that the way I had it hooked up was a bit clunky - and you have to know your dataset when loading it in. So before I messed with it too much more, I wondered if we wanted to approach it differently.

**Comment by rwdudley-usgs** [comment]
Trying to test this to close the issue. @ldecicco-USGS, I've done a pull from master and says it's up-to-date. I still can't get HUCs to map using a HUC dataset. I also notice bugs you already fixed (below) and yet executing an upstream pull says I'm up-to-date? Advice? (and `git push` says up-to-date as well)

```
  if (input$unitTypeHUC == TRUE){
    unit.type=="huc"
  }else{
    unit.type=="county"
  }
```

**Comment by ldecicco-USGS** [comment]
The 2 steps are in the git shell:

```
git pull upstream master
```

Then in RStudio click the Build & Reload button. Try both again, then let me know if you still have errors. I see it's correct on gitHub, so you should be getting it:

https://github.com/USGS-R/wateRuse/blob/master/inst/shiny/mapData.R#L12

**Comment by rwdudley-usgs** [comment]
Strange. I still see those lines in `mapData.R` unchanged.
I've issued `git pull upstream master` (says up-to-date), then `git push` (says up to date) and the Build & Reload. I've also tried `git pull upstream master`, Build & Reload, then `git push` with same result.

I think I'm off on my own branch somewhere? Git says I'm 2 commits ahead. Anyway for me to back out of the commits?

**Comment by ldecicco-USGS** [comment]
Unless there's something you've changed recently that you don't want to lose, you can try this...it will reset your local files to the upstream:

```
git reset --hard upstream/master  
git push origin master --force 
```

**Comment by jshourds-usgs** [comment]
This is a goodie-- adding to documentation ^^  Thanks!

**Comment by ldecicco-USGS** [comment]
Be careful with it though!!!!!! Should always be a last resort.

**Comment by rwdudley-usgs** [comment]
That worked. Out of the woods and back between the guard rails.

**Comment by rwdudley-usgs** [comment]
Ok, I tested the HUC functionality in Shiny and it looks good. We could edit the title to say that works with county and HUC data now instead of only county data.

**Comment by jshourds-usgs** [comment]
Krissi Linsey (NY Water Use Specialist and part of the national water use team) and I spoke last week and she mentioned she has the latest and greatest WBD dataset for HUC coverages that she has been using and she would be more than willing to share with us.  I had shared some jpegs of choropleths showing what we were working on and she thinks her WBD dataset may have a finer scale and possibly better HUC and county linework?  I wasn't sure if the work @rwdudley-usgs and @dblodgett-usgs had done was using this "latest and greatest" dataset and if perhaps lines got "blurred out" because of the polygon work you were doing or because of the changing HUCs/counties issue or you chose to use a different dataset for another reason? Krissi was also familiar with changing HUCs/counties and similarly had some hair-pulling-out experiences in the work she was doing.  She also came to the "sort of solution" of using the latest linework and trying to retrofit historical data to the new HUC and county lines and numbers....  So I'll just put that out there-- Krissi's email address is kslinsey@usgs.gov

**Comment by mamaupin-usgs** [comment]
Thanks  Jen, Yes, Krissi Linsey and Nancy Barber (nlbarber@usgs.gov) are
_the_ two key people that work with the Census population data that we tag
to shapefiles (county/huc) for a compilation year.  They have lost a lot of
hair, I'm sure, working with these datasets.  We should collaborate with
their work. Dave may have already reached out to Nancy.  If we align our
work with them, we close the loop because their population data is the
official county and huc lists and shapefiles that go into AWUDS.

On Fri, Jun 17, 2016 at 10:04 AM, jshourds-usgs notifications@github.com
wrote:

> Krissi Linsey (NY Water Use Specialist and part of the national water use
> team) and I spoke last week and she mentioned she has the latest and
> greatest WBD dataset for HUC coverages that she has been using and she
> would be more than willing to share with us. I had shared some jpegs of
> choropleths showing what we were working on and she thinks her WBD dataset
> may have a finer scale and possibly better HUC and county linework? I
> wasn't sure if the work @rwdudley-usgs https://github.com/rwdudley-usgs
> and @dblodgett-usgs https://github.com/dblodgett-usgs had done was
> using this "latest and greatest" dataset and if perhaps lines got "blurred
> out" because of the polygon work you were doing or because of the changing
> HUCs/counties issue or you chose to use a different dataset for another
> reason? Krissi was also familiar with changing HUCs/counties and similarly
> had some hair-pulling-out experiences in the work she was doing. She also
> came to the "sort of solution" of using the latest linework and trying to
> retrofit historical data to the new HUC and county lines and numbers.... So
> I'll just put that out there-- Krissi's email address is kslinsey@usgs.gov
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-226810414, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLb2K-Sez4_eBCisD-AQzbBTbbD0Shks5qMsV3gaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by jshourds-usgs** [comment]
Exactly. Thanks, Molly.  And as always, thanks for your super clarifying wording-- makes what I was trying to add make sense!  Dave and Rob and Laura may already be using those datasets that Krissi and Nancy are using but can't hurt to double check to make sure we are all using the same supporting data for the same water use data and pulling the same hair out over the same frustrating changing boundaries! ha!

**Comment by ldecicco-USGS** [comment]
I'm going to assign this to @dblodgett-usgs because this falls 100% into his area of expertise on the project vs. mine.

**Comment by dblodgett-usgs** [comment]
We certainly need to compile a set of historical HUC08s for AWUDS to join up to. I used HUCs from June 2015 that should be pretty much the latest at a HUC08 scale. 

Are there specific ones we've identified that are problematic?

**Comment by mamaupin-usgs** [comment]
Yes, and I've included Nancy and Krissi in this email. Krissi has a list
that of changing/problematic HUCs that she's building as she works through
the US computing the HUC8 populations for AWUDS. This list will be part of
the release notes for AWUDS 3.1 coming out very soon.

She has the list of 2015 counties too and (I'm sure) the changes from 2010.

As far as getting historical lists and shapefiles for HUC8 and counties,
that's squarely in Nancy/Krissi and Curtis Price's arena.

On Tue, Jun 21, 2016 at 12:37 AM, David Blodgett notifications@github.com
wrote:

> We certainly need to compile a set of historical HUC08s for AWUDS to join
> up to. I used HUCs from June 2015 that should be pretty much the latest at
> a HUC08 scale.
> 
> Are there specific ones we've identified that are problematic?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-227353005, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLb3rnWF56q4_NLXvywjTiHDY6YM3rks5qN4argaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by mamaupin-usgs** [comment]
Dave, I'm thinking there are really two main sets of HUC08 datasets you
need to worry about, the ones published in the National Atlas (1:2M) and
the current WBD. I know there have been a lot of HUC12 changes.

Curtis Price http://profile.usgs.gov/cprice / USGS ~ science for a
changing world http://www.usgs.gov/ / @cpriceusgs

**Comment by mamaupin-usgs** [comment]
We've done a mix of rolling updates to the AWUDS county and HUC reference
lists, and (especially for the HUCs), more systematic checks associated
with major AWUDS releases and compilation years. The trouble with HUCs is
that the dates of the changes aren't always well tracked in the WBD, though
Krissi discovered a set of metadata items in the last round that documented
some of them. One of those tasks I need to complete is getting 'frozen'
county and HUC8 coverages into ScienceBase or the Water node that match our
AWUDS compilation datasets.   Maybe that will go on next year's work plan...

Dave, the list of the HUC8 changes we made for the AWUDS 3.x release (the
version for the 2015 compilation) is best pulled from the release notes.
There were major changes in HUC8s in Alaska (sigh) and Maine. There's a
change affecting only 1 or 2 HUC8s in NY, but it shifts the watershed to a
different HUC2. There are other deletes and adds scattered around other
States.

The release notes are here. Note that there are some corresponding HUC4
changes that haven't been done yet and aren't listed:
http://nwis.usgs.gov/IT/awuds/Awuds_3_0/awuds_release_notes_3_0.html#ReflistChanges
Note also that there are adds, complete deletes (was never valid in that
State, any data associated with it was moved to the correct HUC), and
'deletes' that turn off the code as of a certain year, but the code is
valid in older datasets.

I'm afraid the current WBD will rarely match an AWUDS dataset until the WBD
stabilizes. That is hopefully getting closer, but the list of changes we
found since the last compilation is a little depressing.

 Nancy

## 

Nancy L. Barber
Hydrologist, USGS South Atlantic Water Science Center
1770 Corporate Drive, Suite 500
Norcross, Georgia 30093
678-924-6627 (office)
678-832-9246 (cell)

**Comment by dblodgett-usgs** [comment]
Isn't the functionality we want for the purpose of comparison? If it's comparing the same watershed from one compilation to the next that we're after, the HUCs that have changed aren't really needed... right? 

If that's the case then I think we are ok just using a coverage that matches close enough and moving on till we have a more comprehensive set of HUCs that match through time. 

It would be a lot of work to try and reconcile things... I know we don't have the bandwidth to do it right in the super near term. Hopefully we can make do with a recent snapshot of HUCs or some other national dataset someone already has an move forward?

**Comment by mamaupin-usgs** [comment]
Yes, comparison between years is one of the functions needed, but so is
simply displaying the data with it's spatial extent (whatever year you're
displaying).  I  understand the level of work is significant to make this
happen, but beyond this utility (R) we'll need to have spatial datasets
that correlate with the tabular data. I think it's wise to keep the task on
our radar but prioritize. I'm not against using the current WBD/County
shapefiles and work from there.

On Wed, Jun 22, 2016 at 2:56 AM, David Blodgett notifications@github.com
wrote:

> Isn't the functionality we want for the purpose of comparison? If it's
> comparing the same watershed from one compilation to the next that we're
> after, the HUCs that have changed aren't really needed... right?
> 
> If that's the case then I think we are ok just using a coverage that
> matches close enough and moving on till we have a more comprehensive set of
> HUCs that match through time.
> 
> It would be a lot of work to try and reconcile things... I know we don't
> have the bandwidth to do it right in the super near term. Hopefully we can
> make do with a recent snapshot of HUCs or some other national dataset
> someone already has an move forward?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/USGS-R/wateRuse/issues/72#issuecomment-227683405, or mute
> the thread
> https://github.com/notifications/unsubscribe/ARyLb1KjbG6vNYzs9X1CwCpwKufG782aks5qOPjNgaJpZM4IaicX
> .

## 

Molly A. Maupin
Hydrologist
USGS Idaho Water Science Center
230 Collins Road
Boise, ID  83702

Office: (208) 387-1307
Fax:    (208) 387-1372

mamaupin@usgs.gov

**Comment by dblodgett-usgs** [comment]
100% agreed. Sounds good.
